### PR TITLE
Remove the greedy planner

### PR DESCRIPTION
**Pull request by systay**

### CODE REVIEW

File: community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/NodeHashJoinPlanningIntegrationTest.scala [hunk 1]
```diff
@@ -60,34 +60,6 @@ class NodeHashJoinPlanningIntegrationTest extends CypherFunSuite with LogicalPla
     result should equal(expected)
   }
 
-  test("Should build plans with leaves for both sides if that is requested by using hints") {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/NodeHashJoinPlanningIntegrationTest.scala:63]: Why did this test have to die? Shouldn't we replace it with an IDP version?

File: community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/greedy/CartesianProductTest.scala [hunk 2]
```diff
@@ -1,70 +0,0 @@
...
-import org.mockito.Mockito._
-import org.neo4j.cypher.internal.compiler.v3_0.HardcodedGraphStatistics
-import org.neo4j.cypher.internal.frontend.v3_0.ast.PatternExpression
-import org.neo4j.cypher.internal.compiler.v3_0.planner.logical.Cost
-import org.neo4j.cypher.internal.compiler.v3_0.planner.logical.Metrics.QueryGraphSolverInput
-import org.neo4j.cypher.internal.compiler.v3_0.planner.logical.plans.{CartesianProduct, LogicalPlan}
-import org.neo4j.cypher.internal.compiler.v3_0.planner.{LogicalPlanningTestSupport, QueryGraph}
-import org.neo4j.cypher.internal.frontend.v3_0.test_helpers.CypherFunSuite
-
-class CartesianProductTest extends CypherFunSuite with LogicalPlanningTestSupport {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/greedy/CartesianProductTest.scala:31]: Is this test class specific to testing how the greedy planner found cartesian products? Do we have similar tests elsewhere for how IDP does cartesian products?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CypherPreParser.scala [hunk 3]
```diff
@@ -46,8 +46,8 @@ case object CypherPreParser extends Parser with Base {
 
   def PlannerOption: Rule1[PreParserOption] = rule("planner option") (
       option("planner", "cost") ~ push(CostPlannerOption)
-    | option("planner", "greedy") ~ push(GreedyPlannerOption)
     | option("planner", "rule") ~ push(RulePlannerOption)
+    | option("planner", "greedy") ~ push(GreedyPlannerOption)
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CypherPreParser.scala:50]: Shouldn't this have been removed?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CypherStatementWithOptions.scala [hunk 4]
```diff
@@ -43,6 +43,8 @@ object CypherStatementWithOptions {
           case VersionOption(v) =>
             val newVersion = mergeOption(version, CypherVersion(v), "Can't specify multiple conflicting Cypher versions")
             recurse(tail, newVersion, planner, runtime, updateStrategy, executionMode)
+          case p: PlannerPreParserOption if p.name == "greedy" =>
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CypherStatementWithOptions.scala]: I prefer using `GreedyPlannerOption.name` here to a literal string. But do we really need to parse it still? Is this just to provide a nicer error message than `unknown planner option`?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/CompilerComparisonTest.scala [hunk 5]
```diff
@@ -56,7 +56,7 @@ class CompilerComparisonTest extends ExecutionEngineFunSuite with QueryStatistic
 
   val compilers = Seq[(String, GraphDatabaseService => CypherCompiler)](
     "legacy (rule)" -> legacyCompiler,
-    "ronja (greedy)" -> ronjaCompiler(GreedyPlannerName),
+    "ronja (greedy)" -> ronjaCompiler(IDPPlannerName),
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/CompilerComparisonTest.scala]: This row makes little sense. Remove?

**Comment by Mats-SX** [comment]
I am a bit concerned over the amount of tests that are being removed. Are we sure that all of them were specific to the greedy planner? Why did we have such tests? Is it worthwhile to try to port them to testing IDP instead?

File: community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/greedy/CartesianProductTest.scala (hunk 2 above)
Comment (craigtaverner) [inline community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/greedy/CartesianProductTest.scala:31]: I think the tests in CartesianProductsOrValueJoinsTest and CartesianProductPlanningIntegrationTest seem to cover this for IDP.

File: community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/NodeHashJoinPlanningIntegrationTest.scala (hunk 1 above)
Comment (systay) [inline community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/NodeHashJoinPlanningIntegrationTest.scala:63]: Well... If you look at the test, it's asserting a pretty dumb plan.  The Selection operator on top of the idx-seek-expand (old lines 79 and 80), is doing the exact same thing as the index seek on the RHS (line 82). Not sure why this was being asserted, but IDP correctly does not produce this plan.

I'd say that if it's important to support this pattern (two index seeks for a single pattern relationship), which is entirely possible, we need to teach the planner new tricks.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CypherPreParser.scala (hunk 3 above)
Comment (systay) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CypherPreParser.scala:50]: In Neo4j 3.0, it's still OK to write:
`CYPHER 2.3 planner=greedy <query>`

That's why the concept is still lingering here.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CypherStatementWithOptions.scala (hunk 4 above)
Comment (systay) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/CypherStatementWithOptions.scala]: Yep. For 3.0. But we still want it for the 2.3 compiler as well, right?

**Comment by systay** [comment]
Just went through the change set again. @Mats-SX I don't share your concern. All I see are unit tests removed for the corresponding units that have also been removed.

File: community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/NodeHashJoinPlanningIntegrationTest.scala (hunk 1 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/NodeHashJoinPlanningIntegrationTest.scala:63]: Now that I think of it, this test is weird. The greedy planner only supported one index hint. How does this even run...?

Currently, supplying two index hints for a single pattern relationship will fallback to rule planner, which I guess is fine.

**Comment by Mats-SX** [comment]
Fair enough. The PR got the attention I urged for, and I'm satisfied.

**Pull request merged**